[FLINK-23512][runtime][checkpoint] Check for illegal modifications of JobGraph with partially finished operators

gaoyunhaii opened the pull request:
## What is the purpose of the change

This PR checks the illegal modification when restoring from a checkpoint with partly-finished operators.

**Bookkeeper the finished task state** 

To implement this PR, we need first store the subtask finished state in the checkpoint. We could not directly use null state as flags of finished, since it could not distinguish the finished one with the operators without state and do not chained with other stateful operators. Another possible options might be let task always report an empty state, but we might not want to affect the normal checkpoints without the finished tasks.

The current implementation stores an explicit flag for the finished tasks, but it seems both of the two options is ok to me. 

**Check the illegal modification**

The check is mainly to avoid that for keyed state, after some tasks finished and the corresponding keyed state is discarded, the topology is modified and a new predecessor is added and re-emitted the records with the discarded keys. The currently logic is that:

1. The predecessors of a fully finished vertex must also be fully finished.
2. The predecessors of a partly finished vertex connected via ALL_TO_ALL edges must be fully finished if the vertex is partly finished.
3. The predecessors of a partly finished vertex connected via POINTWISE edges must be partly finished or fully finished if the vertex is partly finished.

The logic of the check is to ensure:
1. If there is no change and only rescaling, the check must be passed.
2. If there are changes to the topology that causes records to the discarded keys, it must be forbid. But the forbidden operations might not indeed cause records to the discarded keys. 

Currently there is only one bad case is that for a graph composed of `source1 -> forward() -> map1` and `source2 -> forword with key -> map2[with keyed state]` and `source2` and `map2` are connected via `DataStreamUtils#reinterpretAsKeyedStream()`, and after all the vertices become partly finished, users create a new graph with `source1 -> forward with key -> map2[with keyed state]`. It is not very easy to judge this situation, however, in considering the probability of this case is extremely low and the API is only for sophisticated users, perhaps we could first only tip this in the document. 

## Brief change log

- e4992b55912beb45273b8555393bbd9709eb7529 including the finished flag for the finished subtask in the checkpoint.
- cafe182574694b1f708cb9e8318fd54d4ef5ce9f checks for the illegal modification when restoring with partly-finished state.


## Verifying this change

This change added tests and can be verified with the added UT.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: **yes**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

dawidwys commented:
Could you please rebase it on current master and check the test failures? Thanks.

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/VertexFinishedStateChecker.java:
```diff
@@ -0,0 +1,216 @@
+ * <ol>
+ *   <li>All the operators inside a JobVertex have the same finished state.
+ *   <li>The predecessors of a fully finished vertex must also be fully finished.
+ *   <li>The processors of a partly finished vertex
```
```suggestion
 *   <li>The successor of a partly finished vertex
```

gaoyunhaii commented:
> Could you please rebase it on current master and check the test failures? Thanks.

No problem~ I rebased the PR and fixed the serializationuid~ Very thanks for the reivew!

dawidwys commented:
First a few questions to the description. Most probably I just don't understand what is written there.
>The processors of a partly finished vertex connected via ALL_TO_ALL edges must be fully finished if the vertex is partly finished.
> The processors of a partly finished vertex connected via POINTWISE edges must be partly finished or fully finished if the vertex is partly finished.

Do you mean successors or predecessors in those cases? I guess you meant predecessors, right? 

> Currently there is only one bad case is that for a graph composed of source1 -> forward() -> map1 and source2 -> forword with key -> map2[with keyed state] and source2 and map2 are connected via DataStreamUtils#reinterpretAsKeyedStream(), and after all the vertices become partly finished, users create a new graph with source1 -> forward with key -> map2[with keyed state].

I don't get this example. Could you try to explain it a bit more? Do you mean a checkpoint is used by a completely different graph? Is the original graph a disjoint graph? I have a hard time understanding it.

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/FinishedOperatorSubtaskState.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public class FinishedOperatorSubtaskState extends OperatorSubtaskState {
+
+    private static final long serialVersionUID = 7206415348825695023L;
```
Just use `1L`. Please take a look at our coding guidelines for explanation.

gaoyunhaii replied:
Got that, thanks for the guidance~

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/FinishedOperatorSubtaskState.java:
```diff
@@ -0,0 +1,52 @@
+ * A specialized {@link OperatorSubtaskState} used to mark the finished subtasks in the snapshot of
+ * this operator.
+ */
+public class FinishedOperatorSubtaskState extends OperatorSubtaskState {
```
Do we need a separate class? Could we just add a flag to the current class?

gaoyunhaii replied:
It would be also ok to directly add a flag and introduce a special static `FINISHED` instance. Users could not add states to this instance since the underlying `EmptyList` does not support `add` operator. This should be simpler and I updated with this option~

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:
```diff
@@ -347,6 +347,7 @@ public CompletedCheckpoint finalizeCheckpoint(
                 }
 
                 fulfillFullyFinishedOperatorStates();
+                fulfillSubtaskStateForPartlyFinishedOperators();
```
`Partly` -> `Partially`

gaoyunhaii replied:
I updated globally~

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java:
```diff
@@ -114,8 +115,14 @@ protected void serializeOperatorState(OperatorState operatorState, DataOutputStr
+                if (entry.getValue().isFinished()) {
+                    // We store a negative index for the finished subtask. In consideration
+                    // of the index 0, the negative index would start from -1.
+                    dos.writeInt(-(entry.getKey() + 1));
```
That's a bit too much magic for my taste :(

It kind of worked for `Operator` where it meant the number of enclosed states. Here I find too complex especially with the offsetting logic. Unfortunately, I think we need to adjust the metadata version.

dawidwys replied:
@pnowojski @StephanEwen What do you think about adjusting the metadata format to include the finished flag?

pnowojski replied:
Is there no other way around this problem? Creating V4 serialiser would be cleaner, but I think I would be also fine with this hack. Unfortunately I don't have much experience with modifying this code.

Regardless I have one concern. Let's assume subtask `1` ends and it's waiting for final checkpoint, checkpoint `42` happens, it's completed, subtask `1` finishes, checkpoints `43`, `44`... are triggered and completed.

1. We need this flag only for that checkpoint `42`. Previous checkpoints would be still running, checkpoints `43`, and the following ones (assuming we didn't loose `notifyCheckpointComplete()` call) wouldn't contain finished state. Right?
2. And the problem is that if we have even single finished subtasks we have to disallow job graph modifications in front of that subtask, right?
3. In that case, is it enough to write this flag per subtask? What if in checkpoint `44` there was no subtask that was waiting for the final checkpoint? Subtask `1` is long finished. Will we write down the metadata for that subtask `1`, and will we write down this negative number/flag?

gaoyunhaii replied:
Hi @pnowojski ~ Sorry I might not fully understand the issue, but for one thing, I think we would only mark subtask as `isFinished = true` in the checkpoint if it is indeed FINISHED when the checkpoint is taken. This is because if it has called `operator.finish()` and is waiting for the final checkpoint, in the following checkpoint (namely `42` ~ `44`), we should still be able to snapshot its state and thus in this case we would not have the issue of discarding part of the keyed state~ ?

Then in the checkpoint, we might have
1. If when taking the checkpoint some subtasks of an operator is finished, we would save a finished flag (in some way) for these subtasks.
2. If all the subtasks of an operator is finished, we would directly save a finished flag for the whole operator.

Since the FINISHED state of a task won't be reverted, thus in all the following checkpoints, it would also be marked as FINISHED~?

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/VertexFinishedStateChecker.java:
```diff
@@ -0,0 +1,216 @@
+            VertexFinishedState vertexFinishedState = verticesFinishedCache.getOrUpdate(vertex);
+
+            if (vertexFinishedState == VertexFinishedState.FULLY_FINISHED) {
+                checkProcessorsOfFullyFinishedVertex(vertex, verticesFinishedCache);
```
Could you check the naming? What does the `processors` mean?

gaoyunhaii replied:
Yes, it should be `predecessors`, I changed it globally~

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/VertexFinishedStateChecker.java:
```diff
@@ -0,0 +1,216 @@
+                    vertex.getOperatorIDs().stream()
+                            .map(idPair -> checkOperatorFinishedStatus(operatorStates, idPair))
+                            .collect(Collectors.toSet());
+            if (operatorFinishedStates.size() > 1) {
```
nit: `!= 1`?

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/VertexFinishedStateChecker.java:
```diff
@@ -0,0 +1,216 @@
+            predecessorDistribution.compute(
+                    jobEdge.getSource().getProducer().getID(),
+                    (k, v) ->
+                            v == DistributionPattern.ALL_TO_ALL
```
Is it for cases when a vertex is connected twice to the current vertex? If so, could you add such a comment here?

gaoyunhaii replied:
I updated the comments above these lines to be more specific~

dawidwys commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorRestoringTest.java:
```diff
@@ -1213,31 +1213,153 @@ public void testRestoringPartiallyFinishedChainsFails() throws Exception {
     }
 
     @Test
     public void testAddingRunningOperatorBeforeFinishedOneFails() throws Exception {
```
Could we also have a test case for the situation that a single operator is connected via two different distribution patterns? That is quite an uncommon and very specific scenario that in my mind is worth checking.

gaoyunhaii replied:
A test should be needed. I added the test~

gaoyunhaii commented:
Hi Dawid, 

> Do you mean successors or predecessors in those cases? I guess you meant predecessors, right?

It should indeed be the predecessors. I'll check the name overall and change all of them to the predecessors, sorry I initially think processors also means something like predecessors. 

> I don't get this example. Could you try to explain it a bit more? Do you mean a checkpoint is used by a completely different graph? Is the original graph a disjoint graph? I have a hard time understanding it.

Yes, in this example it is indeed a disjoint graph, to be detail:

The main issue here is that we want to avoid if a part of keyed state is finished and discarded, then after failover, we would like to have no new records with these keys. Suppose we have a job composed of two separate parts: 
```
source1 = env.addSource(...)
map1 = source1.forward().map(xx)

source2 = env.addSource(...)
process2 = source2.keyBy().process(xx)
```

Suppose in the first run, `process2` is partly finished, then `source2` must be fully finished since `keyBy` here is `all to all`.  Suppose also `source` and `map1` is also partly finished in the same checkpoint.

After re-run users may modify the graph, There might be several situations:

1. Users add a new operator like
```
source1 = env.addSource(...)
map1 = source1.forward().map(xx)

source3 = env.addSource(...)
process2 = source3.keyBy().process(xx)
```

Then `source3` would be ALL_RUNNING and we could check this situation.

2. users use `source1` to replace the `source2`
```
source1 = env.addSource(...)
process2 = source3.keyBy().process(xx)
```

We could also check this situation since `source1` is partly finished and `keyBy` is all-to-all.

3. users use `DataStream#reinterpretAsKeyedStream()` and use `source1` to replace `source2`
```
source1 = env.addSource(...)
process2 = DataStreamUtils.reinterpretAsKeyedStream(source1, keySelector).process(xx)
```

then in the new graph, `source1` is partly finished, it is connected to `process2` via `forward` edge (POINTWISE). Thus it is not easy to check this situation. 
```

StephanEwen commented:
Thanks for the discussion and digging out the corner cases to check. 
I was wondering is a very simple rule could actually be good enough here:

*You can only change the JobGraph operator chains from a checkpoint if it has no partially finished tasks.*

That would be both simple to implement, simple to explain to users, and would prevent all the corner cases we talked about here. The question is: Is it too restrictive, will it make some common case impossible or in-practical?

_What use cases to we need to support for changing the JobGraph with finished tasks?_
 
  1. Rescaling due to lack of resources and in reactive mode.
  
     This doesn't change the JobGraph structure, but only the parallelism of operators (including possibly operators that are partially finished). @gaoyunhaii is it guaranteed that this still works?
  
  2. Upgrading an application after "stop-with-savepoint" (maybe "stop-with-checkpoint" in the future).
  
      This should be fine, because the final checkpoint should have all tasks finished.
      @pnowojski Can we double check this is also guaranteed with Unaligned Checkpoints?
  
  3. For application joining some bounded inputs and some unbounded inputs, the application often runs the majority of the time after the bounded input finished. They should still be modifiable then.
  
      This should also be fine because in that case the bounded source and the bounded stream sub-flow have only a small time window in which they are partially finished. After that, for they are permanently finished.

_What use cases would be not support?_

  - Upgrades in the middle of a stopping procedure, which has partially completed.
  - Upgrades in the small window while a bounded join input has partially completed (see point (3) above).

Maybe I am overlooking some important use cases here. But if the above is correct, could this really simplify our code and also avoid some unexpected surprises?

Yun Gao mentioned an interesting case with `reinterpretAsKeyedStream()`. Such cases where users make some non-standard assumptions always exist and confuse the users. It would be more explicit for users if they could not restore their checkpoint with changed JobGraph (re-scaling would still work, though) but would get an explicit error and need to bring the job into stable state again.

StephanEwen reviewed (changes requested):
Thanks for the work on this.

## Operator ID handling.

I am unsure about the handling of Operator IDs, in this code and also in previously committed code.

The code only works with the generated Operator ID, and ignores the user-defined operator IDs. I am not sure this is correct, I think we need to register finished state for both generated and user-defined IDs, because on that level, we don't know under which ID the operator will communicate its state. @gaoyunhaii have you looked into this?

To test that, we need a program that uses also user-defined UIDs in the DataStream program. Maybe even a program with two sources and one has explicit user-defined Operator IDs, the other one has auto-generated UIDs.

_Side note: We should refactor this at some point. The user-defined versus generated IDs should be resolved in the API (stream graph generator), before it reaches the runtime layer. On the runtime layer, there should be one Operator ID. I think the current code here has something to do with backwards compatibility back from when we didn't have explicit IDs or when checkpoints didn't track state per operator but per task. /cc @zentol _

## Encapsulating the finished Task logic better.

Currently, the finished task logic is all over the place, and largely in the `CheckpointCoordinator` and in the `PendingCheckpoint`.

I think for a good design, the `PendingCheckpoint` should not need to be aware of `ExecutionJobVertex` and iterate over the status or implement the logic to check for finished state.

That logic belongs into a separate component, probably into the `CheckpointPlan`.
I think the `CheckpointPlan` should have a single method that provides the status for finished tasks:
  - `void fillStateForFinishedTasks(Map<OperatorID, OperatorState> operatorStates)`.
This also simplifies all tests, we can test all the different cases against the `CheckpointPlan` without having to instantiate the `PendingCheckpoint`. I think this is a good indication that this would be a better design.

This also helps with the issue mentioned above with the `OperatorID issue (user-defined versus generated). If we adjust this functionality, we only need to adjust the `CheckpointPlan` and not the `PendingCheckpoint` class.

To go a step further, the `PendingCheckpoint` should also not see the full `CheckpointPlan` interface, but a sub-interface, like `CheckpointOperatorStatusProvider` which only has the above-mentioned method. That makes it easier to test and to adjust.

In general, I think the `PendingCheckpoint` should be completely independent of classes like `ExecutionJobVertex`. When we adjust those classes, or rework the Execution Graph and scheduler, we should not be required to touch the `PendingCheckpoint` Along similar lines, 
private final IdentityHashMap<ExecutionJobVertex, Integer> vertexOperatorsFinishedTasksCount;` should be al test changed to a `Map<JobVertexID, Integer>`.

_(Side note: There are other pre-existing places with that problem as well. For example the `private final Map<ExecutionAttemptID, ExecutionVertex> notYetAcknowledgedTasks` should not expose `ExecutionVertex` but a very small sub-interface that only gives access to the parallelism and the other few properties that are needed.)_

## Encoding of finished status in the Checkpoint Metadata

As mentioned in the comment, the encoding trick done my Yun is an ok hack in my opinion, as long as we don't just do it inline in the methods, but factor the status encoding into static methods.

A metadata version upgrade is cleaner, but it is also quite a bit more involved, mostly because of creating tests for the backwards compatibility.

## Typos / Names

There are a few names that don't fully line up.

  - `fullyFinishedOrFinishedOnRestoreVertex` should be `fullyFinishedOrFinishedOnRestoreVertices`. (Plural of Vertex is Vertices). 
  - `isMayHaveFinishedTasks` also sounds a bit off. Maybe `mayHaveFinishedTasks` or `mightHaveFinishedTasks`.

gaoyunhaii commented:
Hi @StephanEwen very thanks for the review!

> You can only change the JobGraph operator chains from a checkpoint if it has no partially finished tasks. 

I also agree that keeping a simple rule would be more easy to implement and maintain, and the rule would work if there are some difficulty in implementing the checks, the main block comes from that it seems we could not easily acquire the original job graph and do the comparison ?

> The code only works with the generated Operator ID, and ignores the user-defined operator IDs. I am not sure this is correct, I think we need to register finished state for both generated and user-defined IDs, because on that level, we don't know under which ID the operator will communicate its state. @gaoyunhaii have you looked into this?

For the`OperatorIDPair`, currently the name might have some misleadings: the `generatedOperatorID`  is based on both uid and the nature order if uid is not set (`StreamGraphHasherV2`), and the `userDefinedOperatorID` is based on the `uidHash` (`StreamGraphUserHashHasher`). The second one should only be used if users forget to set a customized `uid` on the first run, but want to restore from the state, then users could directly specify the `uidHash` and it will be used as the operator id directly. Thus in the currently logic the userDefinedOperatorID is only checked on restoring to query the corresponding state, but for creating new checkpoint, the `generatedOperatorID` is directly used, when storing the finished status we also follow this policy~

> I think for a good design, the PendingCheckpoint should not need to be aware of ExecutionJobVertex and iterate over the status or implement the logic to check for finished state.

I also agree with the refactor on the CheckpointPlan and PendingCheckpoint  and very thanks for the suggestions! I'll try to update the PR according to the suggestions.
 
From my side, we can do one of the following things:

>1.  Bump the metadata format version. This is relatively simple.
> The tests are the most work here. We need to create Snapshots of V3 metadata that we reload with the latest setup to 
> ensure backwards compatibility.
> We can use the trick here, but then we should nit do this inline, but have two explicit static methods that we use, so that > this is explicit:
> - int encodeSubtaskIndex(int subtaskIndex, boolean isFinished)
> - SubtaskAndStatus decodeSubtaskAndStatus(int value) (where SubtaskAndStatus is like an (int, boolean) tuple.)

For the state metadata format, I'll still more tend to the second options that we first do not upgrade the version and extract them to separate methods, we would still be able to consider upgrading the format if we found other requirements later~

StephanEwen commented:
Thanks for the quick reply!

For most of the questions, looks like we have consensus on how to proceed:
  - The OperatorID issue is in fact not an issue, just a name confusion. I see this issue here is created to reduce the confusion: https://issues.apache.org/jira/browse/FLINK-23681
  - For the Checkpoint Metadata, we encode the status in the current format, but factor out that logic into dedicated methods for clarity.
  - We move the logic for adding the finished state to the `CheckpointPlan`. In the future, there is also a possible optimization to cache the set of finished states and only re-build it when the `CheckpointPlan` changes.

For the rules of when to allow the changes, I understand the issue you raised with not having the other JobGraph to compare to. I think we should avoid a way that needs access to the previous job graph.
  - Can we capture this in rules similar to what you have now, but simpler rules?
  - Or can we solve this by passing another flag into the `restoreCheckpoint()` method, similar to the `allowUnusedState` flag. A flag like `allowPartiallyFinishedTasks`, which is only set to `false` if a new job is submitted?
  - 
I would need to think a bit more here, so maybe for the time being, we could either
  - keep the rules you have for now, but try to document it already in the suggested way (no partially finished tasks). As I understand it, your rules are a more special case of that.
  - not have any complex rules there at all: we allow all modifications, but caution users that upgrades with finished tasks are currently not well-defined. That could be an option if we see that the evaluation of the current rules is either fragile or has a too high overhead with high parallelism jobs.

That would be my final suggestion here also: We need to test that the rules whether restore is allowed scale well to Execution Graphs with > 10000 parallelism and around 100 operators. They should not take longer than few milliseconds to execute.

gaoyunhaii commented:
Hi @StephanEwen, very thanks for the suggestions! I'm currently fixing the code, and for the remaining issues:

> For the rules of when to allow the changes, I understand the issue you raised with not having the other JobGraph to > compare to. I think we should avoid a way that needs access to the previous job graph.
> 
> - Can we capture this in rules similar to what you have now, but simpler rules?
> - Or can we solve this by passing another flag into the restoreCheckpoint() method, similar to the allowUnusedState flag. A flag like allowPartiallyFinishedTasks, which is only set to false if a new job is submitted?

- I also agree now we are checking the specific cases for the more general rules, but it currently seems to me we could not further simple the rule since we might have to consider the cartesian product of `[partly finished, fully finished]` * `[pointwise, all_to_all]`.  
- I think we indeed could add a flag to only do the check on the initial restore (`restoreInitialCheckpointIfPresent` and `restoreSavepoint`. We could always benefit from the flag for failover if we actually has the check. But on the other side, we might still need to consider the same issue for the initial restoring ?


> I would need to think a bit more here, so maybe for the time being, we could either
> 
> - keep the rules you have for now, but try to document it already in the suggested way (no partially finished tasks). As I understand it, your rules are a more special case of that.
> - not have any complex rules there at all: we allow all modifications, but caution users that upgrades with finished tasks are currently not well-defined. That could be an option if we see that the evaluation of the current rules is either fragile or has a too high overhead with high parallelism jobs.
> 
> That would be my final suggestion here also: We need to test that the rules whether restore is allowed scale well to Execution Graphs with > 10000 parallelism and around 100 operators. They should not take longer than few milliseconds to execute.

I agree with that we could only promise to the users that we do not support topology change with finished tasks. I'll first run the tests regarding the performance so that we could make the decision based on the tests results~

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorSubtaskState.java:
```diff
@@ -59,6 +60,8 @@
 
+    public static final OperatorSubtaskState FINISHED = new OperatorSubtaskState(true);
+
     private static final long serialVersionUID = -2394696997971923995L;
```
Yes, since users may upgrade from the older version, we might need to keep the compatibility. 

But to avoid unexpected problems I change the implementation to not modify this class, but introduce a new `FinishedOperatorSubtaskState` subclass specially.

gaoyunhaii commented:
Hi @StephanEwen , very thanks for the review and suggestions! I updated the PR:
1. I encapsulate the logic regarding finished state into `CheckpointPlan` now based on the above discussion. Currently we might not be able to only pass the `FinishStateProvider` to the `PendingCheckpointFinishedTaskStateProvider` directly since it also relies on `CheckpointPlan` to acquire the tasks to trigger / commit / ack from `CheckpointPlan`. So I also extract `CheckpointPlan` to be an interface extends `PendingCheckpointFinishedTaskStateProvider`, and pass `CheckpointPlan` into the PendingCheckpoint for decouple and easy tests. Another option is to pass separate trigger / commit / ack lists together with the `PendingCheckpointFinishedTaskStateProvider` to the `PendingCheckpoint` separately, but it seems a bit verbose to me?
2. Currently the `CheckpointPlan` still relies on `ExecutionVertex` and `ExecutionJobVertex` since the implementation relies on some information like operator lists and parallelism. I also agree with that have a dedicated interfaces like `CheckpointableVertex` and `CheckpointableJobVertex` to only expose the required information, but it seems not trivial and we might do this in dedicate issue?
3. I refactored the `MetadMetadataV3Serializer` and add a flag to only do the illegal modification check for the startup restoring.
4. I tested the performance for the validation with the 15 JobVertex, each vertex has 10 operators and parallelism 1000, the running time is roughly 5~7 ms on my laptop (The average of running 20 times). Thus I think the performance might be acceptable ?

StephanEwen commented:
Thanks, for the refactoring and the performance analysis. At a first glance, this looks very promising.
I'll add some comments in a detailed review next.

gaoyunhaii commented:
Very thanks @StephanEwen!

StephanEwen reviewed (approved):
+1 to merge this PR.

The code is good all in all. There are a few changes I would suggest for some names and some testing structure.

If we commit to doing these changes in a follow-up PR, then I'd be fine with merging this PR without those changes.

I'll add the suggestion for these changes in a separate comment.

StephanEwen commented:
## Suggested Follow-up Changes

### Names

  - I was struggling a bit with the name `isFinishedOnRestore`, for example in `TaskStateSnapshot.isFinishedOnRestore()`.

    What does that mean? Will the task be deployed as "finished" on the next restore? Was it already finished during the last restore?

    Can we make this more self-explaining and call it `taskWasDeployedAsFinished`? /cc @pnowojski 

  - `isOperatorsFinished()` as in `TaskStateSnapshot.isOperatorsFinished()` sounds wrong to me, grammar wise. Given that the class is the "Task Snapshot", should we call the method `isTaskFinised()`. Alternatively, we could call it `areAllOperatorsFinished()`. I would suggest the first version, though.

  - `updateNonFinishedOnRestoreOperatorState` sounds a bit clumsy. Can we call this simply `updateOperatorState(...)` for being teh default case?

### Testing

All the tests about checking the illegal JobGraph modifications are written as tests against the `CheckpointCoordinator`, when they could be written just against the `VertexFinishedStateChecker`. That would make the tests more targeted, like only against the actual component that has the logic. That way, we need less test maintenance when the checkpoint coordinator changes later.

It is probably a good idea to have two test against the `Scheduler` that validate that the modification tests happen at the right points. Something like `testJobGraphModificationsAreCheckedForInitialSavepoint()` and `testJobGraphModificationsAreCheckedForInitialCheckpoint()`.

Then we need no dedicated tests against the `CheckpointCoordinator` regarding illegal job upgrades. That makes sense, because handling this is also the responsibilities of the `Scheduler` and the `VertexFinishedStateChecker`. The `CheckpointCoordinator` is only the component that connects the two, and forwards the calls.

### Further separating the concerns of `PendingCheckpoint` and `CheckpointPlan`

The PR makes some good changes into separating these two components better. But I think we can go a bit further.

The guiding thought is: When dealing with a `PendingCheckpoint`, should one be forced to also deal with a `CheckpointPlan` at all? I think not, the `PendingCheckpoint` is purely a tracker of gathered states and how many tasks have acknowledged and how many we are still waiting for. That makes testing, reusability, etc. simple.

That means we should strive to have the `PendingCheckpoint` independent of `CheckpointPlan`. I think we can achieve this with the following steps:

  - Much of the reason to have the `CheckpointPlan` in `PendingCheckpoint` is because the `CheckpointCoordinator` often requires access to the checkpoint plan for a pending checkpoint. And it just uses the `PendingCheckpoint` as a convenient way to store the `CheckpointPlan`.
  
  A cleaner solution would be changing in `CheckpointCoordinator` the `Map<Long, PendingCheckpoint> pendingCheckpoints` to a `Map<Long, Tuple2<CheckpointPlan, PendingCheckpoint>> pendingCheckpoints`. The `CheckpointCoordinator` would keep the two things it frequently needs together in a tuple, rather than storing one in the other and thus expanding the scope/concerns of `PendingCheckpoint`.

  - The above change should allow us to reduce the interface of `PendingCheckpoint.checkpointPlan` to `PendingCheckpointFinishedTaskStateProvider`.

  - The interface `PendingCheckpointFinishedTaskStateProvider` has some methods that are about tracking state, not just about giving access to finished state: `void reportTaskFinishedOnRestore(ExecutionVertex task)` and `void reportTaskHasFinishedOperators(ExecutionVertex task);`.
  
  I am wondering if those two would not be more cleanly handled outside the `PendingCheckpoint` as well. Meaning in the method `CheckpointCoordinator.receiveAcknowledgeMessage()` we handle the reporting of finished operators to the `CheckpointPlan`. The method `PendingCheckpoint.acknowledgeTask()` only deals with the states.

gaoyunhaii commented:
Hi Stephan @StephanEwen  very thanks for the review and suggestions! They are indeed very instructive to me~

1. The initial thought of `isFinishedOnRestore` is to mark the task as "could be marked as finished after restored", but "wasDeployedAsFinished" seems indeed more smooth to me. 
2. `operatorsFinished` is introduced due to we need to distinguish between the point that task has called `finishOperators()` and is waiting for the last checkpoint / savepoint and the point that task is transit to FINISHED on the JM side, thus I think we might not be able to directly use `isTaskFinised` for the first case ? 
3. I also agree we could limit the scope of tests to test `VertexFinishedStateChecker` only to decouple the checker with the `CheckpointCoordinator`~
4. Currently `PendingCheckpoint` indeed depends on `CheckpointPlan` to provide some meta information about the current checkpoint. It is very instructive to me to let the `CheckpointCoordinator` to directly access the `CheckpointPlan` instead of proxying by the `PendingCheckpoint`. In this way we could be able to separate the logic with the finished states and the reported states. I'll have a try with this direction~

I also agree with we could do the code improvement in a dedicated PR. I opened a new issue https://issues.apache.org/jira/browse/FLINK-23815 for this. For the current PR I also fixed the above minor issues and squashed the commits.

gaoyunhaii commented:
@flinkbot run azure

gaoyunhaii commented:
Very thanks @StephanEwen  for the review! I'll merging the code~
